clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -271,9 +289,9 @@ public void poll(long timeout) throws IOException {
                 try {
                     /* complete any connections that have finished their handshake */
                     if (key.isConnectable()) {
-                        channel.finishConnect();
-                        this.connected.add(channel.id());
-                        this.sensors.connectionCreated.record();
+                    	// queue this channel for completion after this loop
+                    	connectableChannels.add(channel);
+                    	continue;
```

Comment:
Thinking a bit more. This is a bit tricky since we probably can't just continue here. For channels like SSL, we need to do the handshake after the socket is connected. Currently, the handshake will be triggered immediately after the connection is established through channel.prepare() and this has to be done in the same poll(). Otherwise, the selector may not be able to select the key again for initiating the SSL handshake. This applies to all those immediately connected keys not coming from the selector.

So, to get around this issue. We can probably create a new KeyIterator that iterates both keys in this.nioSelector.selectedKeys() and those in connectableChannels. The iterator can return a <key, alreadyConnected> tuple. For keys coming from selectedKeys(), alreadyConnected will be false. For keys from connectableChannels, alreadyConnected will be true. Then, we just need to change line 291 to check if (key.isConnectable()) || alreadyConnected) and leave the rest of the code as it is. This way, keys in  connectableChannels will be handled in the same as way as those from selectedKeys().